### ISSUE TITLE
usability issue with `dune exec` when passing command-line arguments

### ISSUE THREAD

**Issue by gasche** [ocaml/dune#9513]
I run into the following problem *all the time*:

```
$ dune exec minigen --exhaustive --depth 3
dune: unknown option '--exhaustive'.
      unknown option '--depth'.
Usage: dune exec [OPTION]… PROG [ARGS]…
Try 'dune exec --help' or 'dune --help' for more information.
```

The correct invocation is not `dune exec minigen --exhaustive --depth 3`, but `dune exec -- minigen --exhaustive --depth 3`.

Could we actually fix `dune exec` to make the obvious way to write this work as expected?

My proposal would be to just decide that all dune-relevant command-line arguments must be passed *before* the first argument that does not start with a dash: `dune exec minigen --debug-load-dir` must now be written `dune exec --debug-load-dir minigen`, and in general `dune exec foo ...` always behaves as `dune exec -- foo ...`.

In theory this could break compatibility for people who want to pass dune arguments in the middle of their `exec` command, but this is unreadable anyway so nobody should do that, and there is an obvious fix, moving the dune arguments earlier, that is backward-compatible with all previous dune version.

**Comment by gasche** [comment]
Note: an alternative would be to always mandate that `dune exec` is followed by `--` before the user-specific command. Then our muscle memory would remember that one *always* has to write `dune exec --` and not `dune exec`, and the problem would also go away. But then we should rewrite the documentation.

**Comment by rgrinberg** [comment]
I recall some discussions around this, and the root of the matter was that this limitation is inherited from cmdliner and there's not much we can do without forking the library.

**Comment by ejgallego** [comment]
Isn't the `--` separator kind of standard in the Unix / GNU world for these kind of commands?

**Comment by gridbugs** [comment]
I'm personally in favour of requiring `--` between arguments to dune and arguments to the exec'd program. It's consistent with  other tools (e.g. `cargo`, `npm`). Also I think the proposed behaviour would be more confusing in some cases, such as `dune exec foo --release`, where I would expect `--release` to be an argument to dune.

**Comment by dbuenzli** [comment]
More context on why this won't change in `Cmdliner` can be found in the issue that @gasche opened in https://github.com/dbuenzli/cmdliner/issues/183

**Comment by gasche** [comment]
I understand the rationale provided here and by @dbuenzli, but then I would like the documentation to consistently use `dune exec -- foo` instead of `dune exec foo`, which it currently does not.

**Comment by ejgallego** [comment]
Indeed I was going to say the same than Gabriel, I think we can improve the documentation, and always use the `--` form; there are no cases where I've ever invoke `dune exec` without the `--`. The man page could also be more explicit.

**Comment by gridbugs** [comment]
Conventionally the `--` separates the launcher's arguments (e.g. dune's arguments) from the launched program's arguments. But the name of the program to launch is an argument to the launcher. Thus `dune exec foo -- --foo-argument1 --foo-argument2` is the more conventional way to write this rather than `dune exec -- foo --foo-argument1 --foo-argument2` (though both will work the same). This is consistent with other tools where the name of the program to run is omitted or implied like cargo, where you would write `cargo run -- --arg1 --arg2` where the arguments are interpreted by the program being built/run rather than to cargo itself.

**Comment by ejgallego** [comment]
Good point @gridbugs , I use `dune exec -- coqc --foo` tho just because my "mental model" matches thinking of the command I want to run as if Dune wasn't there. Sometimes also because cut and paste.
